File: api/Enkl.Api/Controllers/ProjectsController.cs
using System.Text.Json;
using Enkl.Api.Auth;
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/projects")]
public class ProjectsController : ControllerBase
{
    private readonly ProjectService _projects;

    public ProjectsController(ProjectService projects)
    {
        _projects = projects;
    }

    [HttpGet]
    public async Task<IActionResult> GetProjects()
    {
        var userId = User.UserId();
        return Ok(await _projects.GetProjectsForUserAsync(userId));
    }

    [HttpGet("{projectId:guid}")]
    [Authorize(Policy = "ProjectMember")]
    public async Task<IActionResult> GetProject(Guid projectId)
    {
        var detail = await _projects.GetProjectDetailAsync(projectId);
        return detail is null ? NotFound() : Ok(detail);
    }

    // No ProjectMember policy — the project doesn't exist yet, so there's nothing for that policy to
    // check against. Any authenticated user may create a project under their own Organisation.
    [HttpPost]
    public async Task<IActionResult> CreateProject(CreateProjectRequest request)
    {
        var userId = User.UserId();
        var result = await _projects.CreateAsync(userId, request);
        return result is null ? Unauthorized() : Ok(result);
    }

    // Same "no ProjectMember policy" reasoning as CreateProject above — this is the pre-creation
    // uniqueness check for the "New Project" flow, called before a project (and thus a projectId to
    // scope {projectId:guid}/key-availability below to) exists at all.
    [HttpGet("key-availability")]
    public async Task<IActionResult> CheckKeyAvailabilityForCreation([FromQuery] string key)
    {
        return Ok(await _projects.CheckKeyAvailabilityForCreationAsync(User.OrgId(), key));
    }

    [HttpPut("{projectId:guid}")]
    [Authorize(Policy = "ProjectMember")]
    public async Task<IActionResult> UpdateProject(Guid projectId, UpdateProjectRequest request)
    {
        var result = await _projects.UpdateAsync(projectId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{projectId:guid}")]
    [Authorize(Policy = "ProjectMember")]
    public async Task<IActionResult> DeleteProject(Guid projectId)
    {
        return await _projects.DeleteAsync(projectId) ? NoContent() : NotFound();
    }

    // Org-Admin-only: changing a project's key is restricted well above ordinary project editing
    // (ProjectMember above) — see ProjectService.ChangeKeyAsync's own doc comment for why.
    [HttpGet("{projectId:guid}/key-availability")]
    [Authorize(Policy = "OrgAdmin")]
    public async Task<IActionResult> CheckKeyAvailability(Guid projectId, [FromQuery] string key)
    {
        var result = await _projects.CheckKeyAvailabilityAsync(User.OrgId(), projectId, key);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("{projectId:guid}/key")]
    [Authorize(Policy = "OrgAdmin")]
    public async Task<IActionResult> ChangeKey(Guid projectId, ChangeProjectKeyRequest request)
    {
        var result = await _projects.ChangeKeyAsync(User.OrgId(), projectId, request.NewKey);
        return result is null ? NotFound() : Ok(result);
    }

    // Project Administrator capability: "change app settings" — see ProjectAdminAuthorizationHandler.
    [HttpPut("{projectId:guid}/settings")]
    [Authorize(Policy = "ProjectAdmin")]
    public async Task<IActionResult> UpdateSettings(Guid projectId, ProjectSettingsDto request)
    {
        var result = await _projects.UpdateProjectSettingsAsync(projectId, request, User.HasClaim("orgAdmin", "true"));
        return result is null ? NotFound() : Ok(result);
    }

    // Project Administrator capability: "manage workflow" — see ProjectAdminAuthorizationHandler.
    [HttpPut("{projectId:guid}/workflow")]
    [Authorize(Policy = "ProjectAdmin")]
    public async Task<IActionResult> UpdateWorkflow(Guid projectId, [FromBody] JsonElement request)
    {
        var result = await _projects.UpdateProjectWorkflowAsync(projectId, request);
        return result is null ? NotFound() : Ok(result);
    }
}

File: php-api/src/Controllers/ProjectsController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Services\ProjectService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/** Ported from Controllers/ProjectsController.cs. */
final class ProjectsController extends BaseController
{
    private function service(): ProjectService
    {
        return new ProjectService(Database::connection());
    }

    public function listMine(Request $request, Response $response): Response
    {
        return $this->json($response, $this->service()->getProjectsForUser($this->callerUserId($request)));
    }

    public function detail(Request $request, Response $response, array $args): Response
    {
        $detail = $this->service()->getProjectDetail($args['projectId']);
        return $detail === null ? $this->notFound($response) : $this->json($response, $detail);
    }

    // No ProjectMember membership required — the project doesn't exist yet, so there's nothing for
    // that policy to check against. Any authenticated user may create a project under their own
    // Organisation (see routes.php: this route sits outside the {projectId} group).
    public function create(Request $request, Response $response): Response
    {
        $result = $this->service()->create($this->callerUserId($request), $this->body($request));
        return $result === null ? $this->json($response, ['message' => 'Unauthorized.'], 401) : $this->json($response, $result);
    }

    // Same "no ProjectMember membership required" reasoning as create() above — this is the
    // pre-creation uniqueness check for the "New Project" flow, called before a project (and thus a
    // projectId to scope checkKeyAvailability below to) exists at all.
    public function checkKeyAvailabilityForCreation(Request $request, Response $response): Response
    {
        $key = (string) ($request->getQueryParams()['key'] ?? '');
        return $this->json($response, $this->service()->checkKeyAvailabilityForCreation($this->callerOrgId($request), $key));
    }

    public function update(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->update($args['projectId'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function delete(Request $request, Response $response, array $args): Response
    {
        return $this->service()->delete($args['projectId']) ? $this->noContent($response) : $this->notFound($response);
    }

    // Org-Admin-only: changing a project's key is restricted well above ordinary project editing —
    // see ProjectService::changeKey's own doc comment for why. Both routes are gated by
    // OrgAdminMiddleware in routes.php, not just the ProjectMember group these sit alongside.
    public function checkKeyAvailability(Request $request, Response $response, array $args): Response
    {
        $key = (string) ($request->getQueryParams()['key'] ?? '');
        $result = $this->service()->checkKeyAvailability($this->callerOrgId($request), $args['projectId'], $key);
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function changeKey(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $result = $this->service()->changeKey($this->callerOrgId($request), $args['projectId'], (string) ($body['newKey'] ?? ''));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function updateSettings(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->updateSettings($args['projectId'], $this->body($request), $this->callerIsOrgAdmin($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function updateWorkflow(Request $request, Response $response, array $args): Response
    {
        $raw = (string) $request->getBody();
        $workflow = $raw === '' ? null : json_decode($raw, true);
        $result = $this->service()->updateWorkflow($args['projectId'], $workflow);
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }
}
